crates/protocol/src/batch/single.rs

```diff
@@ -0,0 +1,246 @@
+//! This module contains the [SingleBatch] type.
+
+use crate::{starts_with_2718_deposit, BatchValidity, BlockInfo, L2BlockInfo};
+use alloc::vec::Vec;
+use alloy_eips::BlockNumHash;
+use alloy_primitives::{BlockHash, Bytes};
+use alloy_rlp::{RlpDecodable, RlpEncodable};
+use op_alloy_genesis::RollupConfig;
+
+/// Represents a single batch: a single encoded L2 block
+#[derive(Debug, Default, RlpDecodable, RlpEncodable, Clone, PartialEq, Eq)]
+pub struct SingleBatch {
+    /// Block hash of the previous L2 block. `B256::ZERO` if it has not been set by the Batch
+    /// Queue.
+    pub parent_hash: BlockHash,
+    /// The batch epoch number. Same as the first L1 block number in the epoch.
+    pub epoch_num: u64,
+    /// The block hash of the first L1 block in the epoch
+    pub epoch_hash: BlockHash,
+    /// The L2 block timestamp of this batch
+    pub timestamp: u64,
+    /// The L2 block transactions in this batch
+    pub transactions: Vec<Bytes>,
+}
+
+impl SingleBatch {
+    /// If any transactions are empty or deposited transaction types.
+    pub fn has_invalid_transactions(&self) -> bool {
+        self.transactions.iter().any(|tx| tx.0.is_empty() || tx.0[0] == 0x7E)
+    }
+
+    /// Returns the [BlockNumHash] of the batch.
+    pub const fn epoch(&self) -> BlockNumHash {
+        BlockNumHash { number: self.epoch_num, hash: self.epoch_hash }
+    }
+
+    /// Validate the batch timestamp.
+    pub fn check_batch_timestamp(
+        &self,
+        cfg: &RollupConfig,
+        l2_safe_head: L2BlockInfo,
+        inclusion_block: &BlockInfo,
+    ) -> BatchValidity {
+        let next_timestamp = l2_safe_head.block_info.timestamp + cfg.block_time;
+        if self.timestamp > next_timestamp {
+            if cfg.is_holocene_active(inclusion_block.timestamp) {
+                return BatchValidity::Drop;
+            }
+            return BatchValidity::Future;
+        }
+        if self.timestamp < next_timestamp {
+            return if cfg.is_holocene_active(inclusion_block.timestamp) {
+                BatchValidity::Past
+            } else {
+                BatchValidity::Drop
+            };
+        }
+        BatchValidity::Accept
+    }
+
+    /// Checks if the batch is valid.
```

Comment:
nit: would be great to link to the specs if there are up to date ones

Reply:
Great suggestion.